### PR TITLE
fetchCargoTarball: backport to 20.03

### PR DESCRIPTION
**Pull request by bhipple**
Cherry-picked from PR #78501, minus the README.md migration/design doc, which I'm deleting.

This causes no functional change, but does recompile all Rust applications. It's
necessary to get the new fetcher on the 20.03 branch in order to backport
updated Rust packages in the future.

See #79975 for details.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by Mic92 — APPROVED** [review]
Agreed. We should have that.

**Comment by mmahut** [comment]
Looks good to me.

**Review by misuzu — APPROVED** [review]

**Review by worldofpeace — APPROVED** [review]
Out of my league here as far as knowledge of proceedings, but it seems fine to me.
I'll defer merge to those who can do such in an informed manner.

**Review by Ma27 — COMMENTED** [review]
Before we backport this to a release branch, can you please check out if/why this patch breaks custom build commands for crates in the `buildPhase`? (Example: https://hydra.nixos.org/build/112753750, already fixed in ce84052ef78c2e6b94f76e974d294109bab78948 by using the new fetcher).

**Comment by bhipple** [comment]
Ugh, thanks for the catch (and fix on spotifyd).  The systemic fix unfortunately requires another rebuild-the-world for all the legacy Rust fetches, so I sent the PR with the fix to staging.

Fortunately, it's the legacy implementation that's broken, and the new one that works, so a "workaround" is to do the upgrade work . . . details here: https://github.com/NixOS/nixpkgs/issues/79975#issuecomment-586535069

**Comment by FRidh** [comment]
Please use `cherry-pick -x`.

**Comment by bhipple** [comment]
Reset to origin/staging-20.03, then re-cherry-picked the two commits with `-x`.

**Comment by jonringer** [comment]
did you want to cherry-pick the upgrade fixes as part of this?

**Comment by bhipple** [comment]
The second commit includes the bugfix at an infra level, which means the application bugfix PRs are unnecessary to backport.  For example, this was red on master, but is now green, after the second commit was merged to master:

https://github.com/NixOS/nixpkgs/pull/80219
https://hydra.nixos.org/job/nixpkgs/trunk/xidlehook.x86_64-linux

This means we don't need to backport all of the individual application changes, except when there's some real change that's actually needed (like bumping a version, security fix, etc.).

Aside from this PR, the only other change that I'd like to pre-emptively backport is this one, which changes the default cargo fetcher implementation to match master, so that the `cargoSha256` values are (out of the box) equivalent between the staging branch and master branch:
https://github.com/NixOS/nixpkgs/pull/79978

As with all of these, the only purpose of backporting is to make it easier for maintainers to cherry-pick Rust applications, since otherwise they'd have to re-compute the `cargoSha256`.

**Comment by bhipple** [comment]
Per suggestion, let's wait until https://github.com/NixOS/nixpkgs/pull/80262 makes its way through staging and over to master, which fixes custom patches for Cargo.lock files.

As it currently stands, the implementation works fine for anything that has a Cargo.lock file out of the gate, but there are ~20 packages in Nix that use custom crate mutations outside of the `src` attribute.  These work fine on the legacy fetcher, but they need https://github.com/NixOS/nixpkgs/pull/80262 before they can be upgraded cleanly to the newer one.

**Comment by drewrisinger** [comment]
@bhipple, any plans to renew this backport? I have a package that I'd like to be able to build on both ``release-20.03`` and ``nixpkgs-unstable`` (for my ``nur-packages`` repo), but I understand from this and #79975 that isn't possible. However, it should be possible if this is merged, correct?

**Comment by bhipple** [comment]
With 20.09 not that far away I'd say no.  This is a no-op functionally to the end-user, but it's a rebuild-the-rust-world and a ton of sha changes in the release branch, as well as an API change mid-stable release for those consuming the nix exprs directly.

I'd do something like this as a work around:

```
cargoSha256 = if <condition> then X else Y;
```
where `<condition>` is something like switching on `system.stateVersion == 20.03`.

**Comment by Ma27** [comment]
Rather than `system.stateVersion` (which is a NixOS option and therefore rather hard to actually use in NUR) I'd recommend to use `lib.trivial.release` instead. Other than that I fully agree with @bhipple.

**Comment by drewrisinger** [comment]
Thanks. That's effectively what I already did. Just took a little while to
track down this cause.

**Pull request closed without merging**
